### PR TITLE
Added entropy to stats module

### PR DESCRIPTION
**Pull request by czgdp1807**
#### References
The following papers and links were used and will be used to make the necessary changes:-
1. http://www.math.uconn.edu/~kconrad/blurbs/analysis/entropypost.pdf
2. https://www.crmarsh.com/static/pdf/Charles_Marsh_Continuous_Entropy.pdf
3. https://en.wikipedia.org/wiki/Entropy_(information_theory)
4. https://en.wikipedia.org/wiki/Normal_distribution#Maximum_entropy

#### Brief description of what is fixed or changed
Entropy for univariate continuous distributions has been added.

#### Other comments
Existence of methods to calculate entropy were unkown to me until the making of this PR.
I have included the `entropy` after studying the module so far.  

#### Release Notes

* stats
  * Addition of entropy to stats module

### CODE REVIEW

**Review by ethankward — APPROVED** [review]

**Comment by czgdp1807** [comment]
Yeah, apologies, I just wrote that for debugging purposes. 😅

**Comment by czgdp1807** [comment]
@ethankward Kindly review the changes I have added for discrete random variables and finite random variables.

**Review by ethankward — CHANGES_REQUESTED** [review]

File: sympy/stats/drv_types.py
Comment (czgdp1807) [reply sympy/stats/drv_types.py]: @ethankward Removed `_entropy` from `PoissonDistribution`.
Thanks for the feedback.

**Comment by czgdp1807** [comment]
@ethankward Can you please tell me should `entropy` return a `Lambda` or a constant expression as result?
I was thinking of the latter one since entropy is a definite integral. Waiting for your response.

**Comment by czgdp1807** [comment]
@jksuom Please help me with the query in the above comment.

**Comment by jksuom** [comment]
I think that entropy is a constant, not a Lambda. It may contain parameters if the probability distribution depends on parameters.

**Comment by czgdp1807** [comment]
So I think returning a constant is a better option. We can leave on the user to use Lambda wherever required. What do you think?

**Comment by jksuom** [comment]
Lambda is used to return a function, but entropy is not a function.

**Comment by czgdp1807** [comment]
Thanks for the response. I will make the edits.

**Comment by czgdp1807** [comment]
@jksuom @Upabjojr Please review the final changes I have made to this PR. No more additions are to be made from my side.

**Comment by jksuom** [comment]
It seems to me that entropy is the expectation of `-log(pdf(x))`. Would it be possible to use `compute_expectation` that should work for all distributions, continuous and discrete alike?

**Comment by czgdp1807** [comment]
Yeah, @jksuom entropy is the expectation of `-log(pdf(x))` but when using `E(-log(density(X)(x)))` works well for continuous random variables but gives `Piecewise` functions for discrete and finite random variables. So I thought of making separate methods for different types. Though if you still say, that we should use `compute_expectation` for re-usability of code then I will make the necessary changes.
Your response awaited.

**Comment by jksuom** [comment]
For discrete variables, I would have expected a sum, either unevaluated or evaluated. Maybe there should not be both `X` and `x` in the expression.

**Comment by czgdp1807** [comment]
@jksuom I gave a thought to your point and came up with the following idea,
1. We can use `compute_expectation` for calculating the entropy of continuous random variable since it works out pretty well.
2. For discrete random variables we can go for the definition of entropy given [here](http://www.math.uconn.edu/~kconrad/blurbs/analysis/entropypost.pdf).
What do you say?

**Comment by jksuom** [comment]
There are many precomputed entropies but no examples of more general cases like random variables restricted to an interval.

**Comment by czgdp1807** [comment]
@jksuom I cannot understand your point in the latest comment. Can you please elaborate more?

**Comment by jksuom** [comment]
In crv.py, there are 3 methods named `compute_entropy` but only one test for continuous variables (`X`), and even that is precomputed. There seem to be no tests for the `compute_entropy` methods.

**Comment by czgdp1807** [comment]
> @jksuom I gave a thought to your point and came up with the following idea,
> 
> 1. We can use `compute_expectation` for calculating the entropy of continuous random variables since it works out pretty well.
> 2. For discrete random variables we can go for the definition of entropy given [here](http://www.math.uconn.edu/~kconrad/blurbs/analysis/entropypost.pdf).
>    What do you say?

@jksuom Any views on this?

**Comment by jksuom** [comment]
I believe that the entropy in the discrete case is well established (the sum of `-log(p_i)*p_i`). It behaves as expected: It is non-negative and greater uncertainty implies greater entropy.

The continuous case is more difficult to define consistently. The probability distribution function pdf is scale dependent. It can take large values, which make the entropy negative. Apparently, it need not increase with greater uncertainty.
```
>>> X = Normal('X', 0, 1)
>>> H(X).n(2)
1.4
>>> H(X, X > 0).n(2)
1.5
```
I would expect that there would be less uncertainty when the domain is restricted to `X > 0`. However, the entropy seems to grow in the example. Is there a bug somewhere?

**Comment by oscarbenjamin** [comment]
The first should be:
```
In [1]: S(1)/2 * log(2*pi*E)                                                                                                                                  
Out[1]: 
log(2⋅ℯ⋅π)
──────────
    2     

In [2]: _.evalf()                                                                                                                                             
Out[2]: 1.41893853320467
```
So looks correct (for the 2 digits you showed).

The definition of differential entropy (assuming that's what this is calculating) would suggest that the second example should be the same minus `log(2)` since
```
    H0 = - Integral(f(x) * log(f(x)), (x, -oo, oo)) 
    H1 = - Integral(2*f(x) * log(2*f(x)), (x, 0, oo)) 
       = - Integral(f(x) * log(2*f(x)), (x, -oo, oo))
       = - Integral(f(x) * log(f(x)), (x, -oo, oo)) - Integral(f(x) * log(2), (x, -oo, oo))
       = H0 - log(2)
```

So it should be
```
In [3]: S(1)/2 * log(2*pi*E) - log(2)                                                                                                                         
Out[3]: 
          log(2⋅ℯ⋅π)
-log(2) + ──────────
              2     

In [4]: _.evalf()                                                                                                                                             
Out[4]: 0.725791352644727
```

**Comment by czgdp1807** [comment]
@jksuom @oscarbenjamin Thanks for pointing out my mistake. I forgot to add the lower_bound and upper_bound on integrals. I have updated the PR.

File: sympy/stats/crv.py
Comment (jksuom) [inline sympy/stats/crv.py]: Would it be possible to use something like `self.expectation` or `self.compute_expectation` that would automatically set the proper integration limits?

File: sympy/stats/tests/test_rv.py
Comment (jksuom) [inline sympy/stats/tests/test_rv.py]: This is essentially a repetition. Perhaps something like `H(X, X > 1)` would not look too complicated.

File: sympy/stats/crv.py
Comment (czgdp1807) [reply sympy/stats/crv.py]: I don't think this should be done, since it will give too complex results even for simple answers. Instead, to avoid repetition of code, we can implement a method `stats_integrate` which can be used in computing, `expectation`, `entropy`, `mgf`, `cf` alike. It will automatically set the domain.

Comment (jksuom) [reply sympy/stats/crv.py]: In which way would the results be more complicated?

Comment (czgdp1807) [reply sympy/stats/crv.py]: Using compute_expectation works [here](https://github.com/sympy/sympy/pull/15693/files#diff-6f227c2ec704bbb00895cfa85e7f4d03R557) and gives results as expected. However, it in the above method was giving complex and unevaluated Integrals. Thanks for the suggestion. I will update the PR and apologies for commenting without trying the `compute_expectation` as I mentioned above. :-)

**Comment by czgdp1807** [comment]
`@jksuom Please review the changes I have made on your suggestions.

**Comment by jksuom** [comment]
> expectation gives unevaluated sums

Perhaps `.doit()` could be added in such a case.

**Comment by czgdp1807** [comment]
@jksuom 
1. I don't think `.doit()` will make an effect, since there is already one as shown [here](https://github.com/sympy/sympy/blob/5d45c0adc1cbcbfcf9e07fb0bf974a379826e64d/sympy/stats/rv.py#L691)
2. When I try to execute the following code, error, `TypeError: compute_expectation() got an unexpected keyword argument 'evaluate'`. Is this error need to be fixed since, finding expectation of `L > 2` is mathematically correct, so ideally it should be handled, or conditional expressions are to be left out due to some reasons. If it is so, then I will try to fix this in this PR itself.

```
from sympy import *
from sympy.stats import *
# from sympy.stats.tests.test_rv import test_H
# X = Normal('X', 0, 1)
p = symbols('p')
L = Geometric('L', p)
print(E(L, L > 2))
```

**Comment by czgdp1807** [comment]
@jksuom Maybe entropy can be added only to `rv_interface` using the concept of expectation, since, a distribution can be specified without using entropy. So, adding it as a feature would be more good looking than adding it to crv_types, drv_types, frv_types. I will update if you agree.

**Comment by czgdp1807** [comment]
@jksuom @oscarbenjamin I have simplified the code to compute entropy. Please take a look at it and provide your reviews. Read my last comment too.

**Comment by czgdp1807** [comment]
@Upabjojr Please review this. The [failing job](https://travis-ci.org/sympy/sympy/jobs/516159575) is passing on my system.

**Comment by oscarbenjamin** [comment]
The tests are not run directly on the PR but on the master branch after merging the PR in. Other changes on master mean that in combination with your PR that test fails.

Assuming you've never committed/merged anything directly to master you can test this with
1. git fetch upstream
2. git checkout master
3. git rebase upstream/master
4. git checkout -b test_merge_15693
5. git merge entropy
6. bin/test sympy/stats/tests/test_continuous_rv.py -k test_benini

**Comment by czgdp1807** [comment]
@jksuom Please provide reviews on [this](https://github.com/sympy/sympy/pull/15693#issuecomment-477902372) and the changes I have made to the code.

**Comment by jksuom** [comment]
I would like to see the entropy of a discrete random variable `X` defined as the expectation of another random variable `-log(p)` where `p` is the probability distribution of `X`. Then `p` is also a random variable (with values between 0 and 1). It may be constant (as in the case of a dice) but, in general, not expressible as a function of `X`. I think that there should be a method for creating this random variable `p` from `X`. Then the definition of entropy would be easy to write down.

**Comment by czgdp1807** [comment]
@jksuom I have done something similar to your suggestion. Please take a look.
In [this](https://github.com/sympy/sympy/pull/15693/files#diff-f72fbde05ac7d76ac18a349b68076190R118) line, `pdf(expr)` is a random variable. In other words, it denotes, `P(X)`, where X is a random variable and that line calculates, `E[-log(P(X))]`. It's giving the desired results correctly.

**Comment by jksuom** [comment]
>  this applies to any random variable, i.e., both continuous and discrete.

Is there an example or a test for a discrete variable?

**Comment by czgdp1807** [comment]
Yes, I have included an example in the docs and a test in the test file.

**Comment by czgdp1807** [comment]
@jksuom Please provide your reviews, all tests are passing.

**Comment by jksuom** [comment]
Can you add a few more tests? For instance, a discrete variable and one where pdf is a dict (if those exist).

**Comment by czgdp1807** [comment]
@jksuom I have added a test for `Geometric` random variable. `assert simplify(H(D, D > 2)) == log(2)` tests for the random variable with pdf as `dict`.

**Comment by czgdp1807** [comment]
@jksuom Any more changes needed to be made? Please let me know.

File: sympy/stats/tests/test_rv.py
Comment (jksuom) [inline sympy/stats/tests/test_rv.py]: It should not be necessary to use strings for comparison if you write `S(1)/2` instead of `1/2`.

File: sympy/stats/tests/test_rv.py
Comment (jksuom) [inline sympy/stats/tests/test_rv.py]: Why `simplify`?

Comment (czgdp1807) [reply sympy/stats/tests/test_rv.py]: I have removed it. The tests are passing without it on my system. I just used it for safety. Thanks for asking. It wasn't required here.

File: sympy/stats/tests/test_rv.py
Comment (czgdp1807) [reply sympy/stats/tests/test_rv.py]: Thanks for the suggestion. I have removed the string comparison.

**Comment by czgdp1807** [comment]
@jksuom I have made the necessary changes and all checks are passing. Please suggest more changes, if any?

**Comment by jksuom** [comment]
I think this looks fairly good now. The entropy is computed as an expectation in a single function. I find it a little disturbing that a pdf given as a dict has to be treated separately. Maybe that could also be transformed to a random variable, though not necessarily in this PR.

@Upabjojr What do you think?

**Comment by czgdp1807** [comment]
@jksuom Infact that's disturbing for me too. I thought a lot for the unification of the entropy calculation logic for distributions of PDF as a dict. I tried to convert dict into Piecewise. If there exists any method/function which can convert dict into Piecewise then the whole logic can be unified.

**Comment by jksuom** [comment]
I think this is good to go. Thanks.

**Pull request merged**
